backend-ktor/src/main/kotlin/me/zwsmith/plugins/database/RecipeService.kt

```diff
@@ -0,0 +1,135 @@
+package me.zwsmith.plugins.database
+
+import kotlinx.coroutines.Dispatchers
+import me.zwsmith.plugins.Ingredient
+import me.zwsmith.plugins.QuantifiedIngredient
+import me.zwsmith.plugins.Recipe
+import me.zwsmith.plugins.recipes.QuantifiedIngredientInput
+import org.jetbrains.exposed.sql.Database
+import org.jetbrains.exposed.sql.SchemaUtils
+import org.jetbrains.exposed.sql.transactions.experimental.newSuspendedTransaction
+import org.jetbrains.exposed.sql.transactions.transaction
+
+class RecipeService(database: Database) {
+    init {
+        transaction(database) {
+            SchemaUtils.create(Recipes, Ingredients, QuantifiedIngredients)
+        }
+    }
+    
+    suspend fun createIngredient(name: String): Ingredient = dbQuery {
+        IngredientDao.new {
+            this.name = name
+        }.toIngredient()
+    }
+    
+    suspend fun findIngredientById(id: Int) = dbQuery {
+        IngredientDao.findById(id)?.toIngredient()
+    }
+    
+    suspend fun updateIngredient(updateData: Ingredient) = dbQuery {
+        IngredientDao.findById(updateData.id)?.apply {
+            name = updateData.name
+        }?.toIngredient()
+    }
+    
+    suspend fun deleteIngredient(id: Int) = dbQuery {
+        val ingredient = IngredientDao.findById(id)
+        if(ingredient == null) {
+            throw IllegalArgumentException("No Ingredient found with ID $id")
+        } else {
+            ingredient.delete()
+        }
+    }
+    
+    suspend fun createRecipe(
+        name: String,
+        description: String,
+        ingredientInputs: List<QuantifiedIngredientInput>
+    ): Recipe {
+        return dbQuery {
+            val recipe = RecipeDao.new {
+                this.name = name
+                this.description = description
+            }
+            
+            // Associate QuantifiedIngredients with the new Recipe
+            ingredientInputs.forEach { input ->
+                val ingredient = IngredientDao
+                    .find { Ingredients.name eq input.ingredient.name }
```

Comment:
is `ingredient.name` uniqueness enforced?

Reply:
Oh woops not in this setup. I need to update the table definition.

Reply:
I ask because if uniqueness is not enforced, you may end up with the wrong ingredient that has the same name (contrived edge case but its best to make invalid states unrepresentable)